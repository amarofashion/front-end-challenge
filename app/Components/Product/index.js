import React from 'react'

export default class Product extends React.Component {
  constructor () {
    super()
  }

  /**
   * Return formatted string for prices.
   */
  formattedPrices () {
    if (this.props.product.on_sale) {
      return (
        <div className="prices">
          <p className="sale"><del>De: {this.props.product.regular_price}</del></p>
          <p>Por: {this.props.product.actual_price}</p>
          <p><small>{this.props.product.installments}</small></p>
        </div>
      )
    }

    return (
      <div className="prices">
        <p>{this.props.product.actual_price}</p>
        <p><small>{this.props.product.installments}</small></p>
      </div>
    )
  }

  /**
   * Return only availables options.
   *
   * @param  {Object} product
   * @return {Array}
   */
  options () {
    return this.props.product.sizes.filter(size => size.available)
  }

  render () {
    return (
      <div className="product-card">
        <figure className="product-cover">
          <img className="image" src={this.props.product.image} alt={this.props.product.name} />
        </figure>

        <div className="product-info">
          <h2 className="title">{this.props.product.name}</h2>

          <div className="product-meta">
            {this.formattedPrices()}
            {
              this.options().map(option => (
                <div key={option.sku} className="option">
                  <input type="checkbox" id={option.sku} hidden />
                  <label htmlFor={option.sku}>{option.size}</label>
                </div>
              ))
            }
          </div>
        </div>

        <button className="to-cart" onClick={() => this.props.cart(this.props.product)}>
          <span>adicionar ao carrinho</span>
          <span className="fa fa-cart-plus fa-2x"></span>
        </button>
      </div>
    )
  }
}
