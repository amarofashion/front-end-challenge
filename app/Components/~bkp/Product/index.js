import React from 'react'

function formatPrice (product) {
  if (product.on_sale) {
    return (
      <div className="prices">
        <p className="sale"><del>De: {product.regular_price}</del></p>
        <p>Por: {product.actual_price}</p>
        <p><small>{product.installments}</small></p>
      </div>
    )
  }

  return (
    <div className="prices">
      <p>{product.actual_price}</p>
      <p><small>{product.installments}</small></p>
    </div>
  )
}

function filterOptions (product) {
  return product.sizes.filter(size => size.available)
}

function handleOption (option, product) {
  if (!product.cart) {
    product.cart = []
  }

  return product.cart = [...product.cart, option]
}

const Product = (product, handleCart) => (
  <div key={product.code_color} className="product-card">
    <figure className="product-cover">
      <img className="image" src={product.image} alt={product.name} />
    </figure>

    <div className="product-info">
      <h2 className="title">{product.name}</h2>

      <div className="product-meta">
        {formatPrice(product)}
        {
          filterOptions(product).map(option => (
            <div key={option.sku} className="option">
              <input
                className="options"
                type="radio"
                id={option.sku}
                name={`product-${product.style}`}
                onChange={() => handleOption(option, product)}
                hidden
              />
              <label htmlFor={option.sku}>{option.size}</label>
            </div>
          ))
        }
      </div>
    </div>

    <button className="to-cart" onClick={event => handleCart(event, product)}>
      <span>adicionar ao carrinho</span>
      <span className="fa fa-cart-plus fa-2x"></span>
    </button>
  </div>
)

export default Product
