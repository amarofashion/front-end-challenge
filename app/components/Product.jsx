import React, { Component } from 'react';
import {Button} from 'react-bootstrap';

import utils from '../utils/utils.jsx';

class Product extends Component {
    constructor(props) {
        super(props);
    }   

    render() {
        var {name, image, regular_price, actual_price, sizes} = this.props;

        // Render prices according to the values of the regular_price and actual_price variables
        var renderPrices = () => {
            if (regular_price === actual_price) {
                return (
                    <div className="col-xs-12 col-prices-not-discounted">
                        <span className="regular-price">{regular_price}</span>
                    </div>
                );
            } else {
                return (
                    <div className="col-xs-12 col-prices-discounted">
                        <div className="col-xs-6">
                            <span className="regular-price">{regular_price}</span>
                        </div>
                        <div className="col-xs-6">
                            <span className="actual-price">{actual_price}</span>
                        </div>                        
                    </div>
                );
            }
        }        

        // Render a span for each size of the product with adequate classes
        var renderSizes = () => {
            let sizeSpanClass;
            return sizes.map((size, index) => {
                sizeSpanClass = size.available ? "span-size span-size-available" : "span-size span-size-not-available";
                return <span key={index} className={sizeSpanClass}>{size.size}</span>
            })
        }

        return (
            <div className="product-container">
                <div className="row">
                    <div className="col-xs-12 col-product-image">
                        <img src={image === '' ? 'http://placehold.it/470x594' : image} alt="" className="img img-responsive"/>
                    </div>
                    <div className="col-xs-12 col-product-name">
                        <p>{utils.capitalizeFirstLetterEachWord(name)}</p>
                    </div>                    
                    {renderPrices()}
                    <div className="col-xs-12 col-sizes">
                        <p>{renderSizes()}</p>
                    </div>
                    <div className="col-xs-12 col-add-to-cart">
                        <Button bsStyle="success" block onClick={() => {this.props.onAddToCart(name, image, actual_price)}}><i className="fa fa-cart-arrow-down" aria-hidden="true"></i> Comprar</Button>
                    </div>
                </div>
            </div>
        );
    }
}

export default Product;