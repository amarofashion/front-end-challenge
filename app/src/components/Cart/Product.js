import React from 'react'
import { connect } from 'react-redux'

import {
  removeProduct,
  changeTotal,
  changeQuantity
} from '../../actions/cartActions'

const mapDispatchToProps = (dispatch => ({
  remove (id) {
    dispatch(removeProduct(id))
    dispatch(changeTotal())
  },

  change (id, value) {
    dispatch(changeQuantity(id, value))
    dispatch(changeTotal())
  }
}))

const Product = props => {
  const { item, remove } = props

  return (
    <section className="product on-cart">
      <figure className="product-cover">
        {
          item.image
            ? (<img src={item.image} alt={item.name}/>)
            : <span className="ion-image"></span>
        }
      </figure>
      <div className="product-meta">
        <div className="product-title">
          <h3>{item.name}</h3>
          <button className="product-remove" onClick={() => remove(item.cid)}>&times;</button>
        </div>

        <div className="product-prices">
          <div className="product-quantity">
            <input type="text" value={item.quantity} />
            <button className="product-increase" onClick={() => props.change(item.cid, 1)}>+</button>
            <button className="product-decrease" onClick={() => props.change(item.cid, -1)}>-</button>
          </div>
          <strong>{item.cart_price.toLocaleString('pt-br', {style: 'currency', currency: 'BRL'})}</strong>
        </div>
      </div>
    </section>
  )
}

export default connect('', mapDispatchToProps)(Product)
