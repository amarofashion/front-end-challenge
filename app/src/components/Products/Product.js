import React from 'react'
import { connect } from 'react-redux'

import { addProduct, changeTotal } from '../../actions/cartActions'

const mapDispatchToProps = (dispatch => ({
  onBuyClick (product) {
    dispatch(addProduct(product))
    dispatch(changeTotal())
  }
}))

const Product = props => {
  const { item } = props

  return (
    <section className="product">
      <figure className="product-cover">
        {
          item.image
            ? (<img src={item.image} alt={item.name}/>)
            : <span className="ion-image"></span>
        }
      </figure>
      <div className="product-meta">
        <h3 className="product-title">{item.name}</h3>
        <div className="product-prices">
          <p><strong>por: {item.actual_price}</strong> ou {item.installments}</p>
          {
            item.on_sale
              ? (<p><small>de: {item.regular_price}</small> <span>({item.discount_percentage})</span></p>)
              : null
          }
        </div>
        <div className="product-sizes">
          {
            item.sizes.filter(size => {
              return size.available
            }).map(({size, sku}) => <a key={sku} href="#">{size}</a>)
          }
        </div>
      </div>
      <button className="product-buy" onClick={() => props.onBuyClick(item)}>
        <span>Comprar</span>
      </button>
    </section>
  )
}

export default connect('', mapDispatchToProps)(Product)
