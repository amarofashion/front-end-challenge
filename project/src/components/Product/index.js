import React from 'react';
import { string, bool, func, shape, arrayOf } from 'prop-types';

import { StyledWrapper } from './style';

const propTypes = {
  image: string.isRequired,
  name: string.isRequired,
  regularPrice: string.isRequired,
  actualPrice: string.isRequired,
  discountPercentage: string.isRequired,
  installments: string.isRequired,
  onSale: bool.isRequired,
  sizes: arrayOf(
    shape({
      available: bool,
      size: string,
      sku: string,
    })
  ).isRequired,
  callback: shape({
    addToCart: func,
  }).isRequired,
};

function Product({
  image,
  name,
  regularPrice,
  onSale,
  actualPrice,
  discountPercentage,
  installments,
  sizes,
  callback: { addToCart },
}) {
  return (
    <StyledWrapper onClick={() => addToCart()} onSale={onSale}>
      <div className='content'>
        <div className='image'>
          {onSale && (
            <div className='sale'>
              <p className='title'>SALE</p>
              <p className='percentage'>{discountPercentage}</p>
            </div>
          )}

          <img src={image} alt={`Foto do produto ${name}`} />
        </div>

        <div className='bottomContent'>
          <p className='name'>{name}</p>

          {onSale
            ? [
                <p className='regularPrice prices'>{regularPrice}</p>,
                <p className='actualPrice prices'>{actualPrice}</p>,
                <p className='installments prices'>{installments}</p>,
              ]
            : [
                <p className='regularPrice prices'>{regularPrice}</p>,
                <p className='installments prices'>{installments}</p>,
              ]}

          <div className='sizes'>
            {sizes.map(
              ({ available, size }) =>
                available && (
                  <div className='size'>
                    <span>{size}</span>
                  </div>
                )
            )}
          </div>
        </div>
      </div>
    </StyledWrapper>
  );
}

Product.propTypes = propTypes;

export default Product;
