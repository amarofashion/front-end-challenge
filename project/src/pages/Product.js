import React from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components';

import colors from '../theme/colors';

const StyledWrapper = styled.div`
  ul {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: stretch;
    justify-content: stretch;
    margin-left: -15px;
    margin-right: -15px;

    li {
      display: flex;
      flex-flow: column;
      align-items: stretch;
      justify-content: stretch;
      width: calc(50% - 30px);
      height: auto;
      margin: 0 15px 65px;

      .content {
        border: 1px solid ${colors.dark};
        border-bottom: none;
      }

      .image {
        position: relative;

        img {
          width: 100%;
          object-fit: cover;
        }
      }

      button {
        padding: 10px 0;
        font-size: 20px;
        text-transform: uppercase;
        text-align: center;
        color: ${colors.light};
        background: ${colors.primary};

        &:hover {
          cursor: pointer;
        }
      }
    }
  }
`;

function Products() {
  const list = useSelector(state => state.products);

  return (
    <StyledWrapper>
      <ul>
        {list.map((product, index) => (
          <li key={index}>
            <div className='content'>
              <div className='image'>
                <img src={product.image} alt='' />
              </div>
              <p>{product.name}</p>
              <p>{product.regular_price}</p>
              <p>{product.on_sale ? 'promoção' : ''}</p>
              <p>{product.on_sale && product.discount_percentage}</p>
              <p>{product.on_sale && product.actual_price}</p>
              <ul>
                {product.sizes.map(item => (
                  <p>{item.size}</p>
                ))}
              </ul>
            </div>
            <button type='button' onClick={() => {}}>
              COMPRAR
            </button>
          </li>
        ))}
      </ul>
    </StyledWrapper>
  );
}

export default Products;
