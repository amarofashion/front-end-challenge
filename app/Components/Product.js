import React from 'react'
import uuid from 'uuid'
import PubSub from 'pubsub-js'

import Icon from './Icon'

import * as CartStorage from '../Utils/CartStorage'
import { handleImage } from '../Utils/Products'

function renderDiscount (discount) {
  if (!discount) {
    return
  }

  return (
    <span className="onsale">{discount}</span>
  )
}

function renderPrice (item) {
  if (item.on_sale) {
    return (
      <div className="product-price">
        <del className="normal">{item.regular_price}</del>
        <p className="ammount">{item.actual_price}</p>
        <small className="installments">{item.installments}</small>
      </div>
    )
  }

  return (
    <div className="product-price">
      <p className="ammount">{item.actual_price}</p>
      <small className="installments">{item.installments}</small>
    </div>
  )
}

function toCart (item) {
  item.quantity = 1
  item.id = uuid()

  CartStorage.add(item)

  let cart = CartStorage.get()

  PubSub.publish('cart', cart)
}

const Product = ({item}) => (
  <div className="card-product">
    {renderDiscount(item.discount_percentage)}

    <figure className="cover">
      <img className="image" src={handleImage(item.image)} alt={item.name} />
    </figure>

    <div className="product-meta">
      <h2 className="title">{item.name}</h2>

      {renderPrice(item)}

      <button className="product-buy" onClick={() => toCart(item)}>
        <span>Adicionaro ao carrinho</span>
        <Icon name="cart" />
      </button>
    </div>
  </div>
)

export default Product
