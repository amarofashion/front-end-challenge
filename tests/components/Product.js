import React from 'react';
import renderer from 'react-test-renderer';

import Product from '../../src/components/Product.jsx';

describe('Product', () => {
	let _Product;
	beforeEach(() => {
		_Product = new Product({});
	});

	test('renderName', () => {
		const name = renderer.create(_Product.renderName({name: 'teste'})).toJSON();
		expect(name).toMatchSnapshot();
	});

	test('getAvailableSizes', () => {
		const sizes = _Product.getAvailableSizes([
		{
			"available": false,
			"size": "PP",
		}, {
			"available": true,
			"size": "P",
		}, {
			"available": true,
			"size": "M",
		}, {
			"available": true,
			"size": "G",
		}, {
			"available": false,
			"size": "GG",
		}
		]);
		expect(sizes).toEqual('P,M,G');
	});

	test('renderSizes null', () => {
		const sizes = _Product.renderSizes({});
		const sizes2 = _Product.renderSizes({sizes: []});

		expect(sizes).toBe(null);
		expect(sizes2).toBe(null);
	});

	test('renderSizes', () => {
		const sizes = renderer.create(_Product.renderSizes({sizes: [
		{
			"available": false,
			"size": "PP",
		}, {
			"available": true,
			"size": "P",
		}, {
			"available": true,
			"size": "M",
		}, {
			"available": true,
			"size": "G",
		}, {
			"available": false,
			"size": "GG",
		}
		]})).toJSON();

		expect(sizes).toMatchSnapshot();
	});

	test('renderImage', () => {
		const image = renderer.create(_Product.renderImage({image: 'url', name: 'nm'})).toJSON();

		expect(image).toMatchSnapshot();
	});

	test('renderCartButton', () => {
		const button = renderer.create(_Product.renderCartButton()).toJSON();

		expect(button).toMatchSnapshot();
	});

	test('toCart', () => {
		let cart = null;
		const _Product = new Product({toCart: item => {
			cart = item;
		}});

		expect(cart).toBeNull();
		_Product.toCart();
		expect(cart).toBeTruthy();
	});

	test('render', () => {
		const _Product = renderer.create(<Product toCart={() => {}} image="url" name="nome" actual_price="R$ 10,00" regular_price="R$ 20,00" discount_percentage="50%" installments="2x R$ 5,00" sizes={[{"available": false, "size": "PP", }, {"available": true, "size": "P", }, {"available": true, "size": "M", }, {"available": true, "size": "G", }, {"available": false, "size": "GG", } ]} />).toJSON();
		expect(_Product).toMatchSnapshot();
	});
});