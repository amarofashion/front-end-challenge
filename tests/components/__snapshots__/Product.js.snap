// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product render 1`] = `
<div
  className="product-box "
  id="pundefined"
>
  <img
    alt="Foto de nome"
    className="image"
    src="url"
  />
  <h2
    className="name"
  >
    nome
  </h2>
  <span
    className="price"
  >
    <span>
      De 
      <del>
        R$ 20,00
      </del>
       
    </span>
    <span>
      por 
      R$ 10,00
       
    </span>
    <em>
      (
      50%
       de desconto) 
    </em>
    <span>
      2x R$ 5,00
    </span>
  </span>
  <span
    className="sizes"
  >
    Tamanhos disponíveis: 
    P,M,G
  </span>
  <button
    className="tocart"
    onClick={[Function]}
    type="button"
  >
    Adicionar ao carrinho
  </button>
</div>
`;

exports[`Product renderCartButton 1`] = `
<button
  className="tocart"
  onClick={[Function]}
  type="button"
>
  Adicionar ao carrinho
</button>
`;

exports[`Product renderImage 1`] = `
<img
  alt="Foto de nm"
  className="image"
  src="url"
/>
`;

exports[`Product renderName 1`] = `
<h2
  className="name"
>
  teste
</h2>
`;

exports[`Product renderSizes 1`] = `
<span
  className="sizes"
>
  Tamanhos disponíveis: 
  P,M,G
</span>
`;
